pkg/ipmi/oem.go

```diff
@@ -0,0 +1,333 @@
+	_FB_OEM_SET_DIMM_INFO = 0x12
+)
+
+type oemData interface {
+	data()
+}
+
+type oemProcProductNameDataParam1 struct {
+	ManufacturerID    [3]uint8
+	Index             uint8
+	ParameterSelector uint8
+	ProductName       [48]byte
+}
+
+func (ProccessorDataParam1 *oemProcProductNameDataParam1) data() {}
+
+type oemProcBasicInfoDataParam2 struct {
+	ManufacturerID        [3]uint8
+	Index                 uint8
+	ParameterSelector     uint8
+	CoreNumber            uint8
+	ThreadNumberLSB       uint8
+	ThreadNumberMSB       uint8
+	ProcessorFrequencyLSB uint8
+	ProcessorFrequencyMSB uint8
+	Revision1             uint8
+	Revision2             uint8
+}
+
+func (ProccessorDataParam2 *oemProcBasicInfoDataParam2) data() {}
+
+type oemDimmLocationDataParam1 struct {
+	ManufacturerID    [3]uint8
+	Index             uint8
+	ParameterSelector uint8
+	DIMMPresent       uint8
+	NodeNumber        uint8
+	ChannelNumber     uint8
+	DIMMNumber        uint8
+}
+
+func (DimmDataParam1 *oemDimmLocationDataParam1) data() {}
+
+type oemDimmTypeDataParam2 struct {
+	ManufacturerID    [3]uint8
+	Index             uint8
+	ParameterSelector uint8
+	DIMMType          uint8
+}
+
+func (DimmDataParam2 *oemDimmTypeDataParam2) data() {}
+
+type oemDimmSpeedDatapParam3 struct {
+	ManufacturerID    [3]uint8
+	Index             uint8
+	ParameterSelector uint8
+	DIMMSpeed         uint16
+	DIMMSize          uint32
+}
+
+func (DimmDataParam3 *oemDimmSpeedDatapParam3) data() {}
+
+type oemDimmModulePartNumberDataParam4 struct {
+	ManufacturerID    [3]uint8
+	Index             uint8
+	ParameterSelector uint8
+	ModulePartNumber  [20]byte
+}
+
+func (DimmDataParam4 *oemDimmModulePartNumberDataParam4) data() {}
+
+type oemDimmModuleSerialNumberDataParam5 struct {
+	ManufacturerID     [3]uint8
+	Index              uint8
+	ParameterSelector  uint8
+	ModuleSerialNumber uint32
+}
+
+func (DimmDataParam5 *oemDimmModuleSerialNumberDataParam5) data() {}
+
+type oemDimmModuleManufacturerIDDataParam6 struct {
+	ManufacturerID          [3]uint8
+	Index                   uint8
+	ParameterSelector       uint8
+	ModuleManufacturerIDLSB uint8
+	ModuleManufacturerIDMSB uint8
+}
+
+func (DimmDataParam6 *oemDimmModuleManufacturerIDDataParam6) data() {}
+
+type BoardManufacturerID struct {
+	ID1 uint8
+	ID2 uint8
+	ID3 uint8
+}
+
+var BoardManufacturer = map[string]BoardManufacturerID{
+	"Wiwynn": {ID1: 0x0,
+		ID2: 0x9c,
+		ID3: 0x9c,
+	},
+}
+
+func (i *IPMI) SendOemIpmiCommand(netfn byte, cmd byte, data oemData) error {
+	req := &req{}
+	req.msg.cmd = netfn
+	req.msg.netfn = cmd
+
+	buf := &bytes.Buffer{}
+
+	if err := binary.Write(buf, binary.LittleEndian, data); err != nil {
+		return err
+	}
+
+	req.msg.data = unsafe.Pointer(&buf.Bytes()[0])
+	req.msg.dataLen = uint16(len(buf.Bytes()))
+
+	_, err := i.sendrecv(req)
+
+	return err
+}
+
+// IpmiOemSetProcessorAndDimmInfo sends the information of proccessors and DIMMs to BMC
+func (i *IPMI) IpmiOemSetProcessorAndDimmInfo(si *smbios.Info) error {
```

Comment:
Is there any reason why you have the processor and dimm in the same function instead of splitting it?

Reply:
I split it into two functions.